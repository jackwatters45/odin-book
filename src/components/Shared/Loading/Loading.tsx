import { styled, keyframes } from "styled-components";

const spin = keyframes`
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
`;

interface SpinnerProps {
	size: number;
	speed: number;
}

const Spinner = styled.div<SpinnerProps>`
	position: absolute;
	margin: auto;
	left: 0;
	right: 0;
	top: 0;
	bottom: 0;

	border: ${({ size }) => size / 4}px solid #f3f3f3;
	border-top: ${({ size }) => size / 4}px solid rgba(55, 55, 55, 0.5);
	border-radius: 50%;
	width: ${({ size }) => size}px;
	height: ${({ size }) => size}px;
	animation: ${spin} ${({ speed }) => speed}s linear infinite;
`;

interface Props {
	className?: string;
	size?: number;
	speed?: number;
}

const Loading = ({ className, size = 50, speed = 1.5 }: Props) => {
	return <Spinner className={className} size={size} speed={speed} />;
};

export default Loading;
