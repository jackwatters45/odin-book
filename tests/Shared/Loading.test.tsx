import { render } from "@testing-library/react";
import { describe, it, expect } from "vitest";
import "jest-styled-components";

import Loading from "../../src/components/Shared/Loading";

describe("Loading Component", () => {
	it("renders without crashing", () => {
		render(<Loading />);
		const spinner = document.querySelector("div");
		expect(spinner).toBeInTheDocument();
	});

	it("applies correct styles based on the size prop", () => {
		const { container } = render(<Loading size={40} />);
		const spinner = container.firstChild as HTMLElement;

		expect(spinner).toHaveStyleRule("width: 40px");
		expect(spinner).toHaveStyleRule("height: 40px");
		expect(spinner).toHaveStyleRule("border: 10px solid #f3f3f3");
		expect(spinner).toHaveStyleRule("border-top: 10px solid rgba(55, 55, 55, 0.5)");
	});

	it("applies correct className based on the className prop", () => {
		const { container } = render(<Loading className="test" />);
		const spinner = container.firstChild as HTMLElement;

		expect(spinner).toHaveClass("test");
	});

	it("has the spin animation", () => {
		const { container } = render(<Loading />);
		const spinner = container.firstChild as HTMLElement;

		const style = window.getComputedStyle(spinner);
		expect(style.animationName).toBeDefined();
	});

	it("matches the snapshot", () => {
		const { asFragment } = render(<Loading />);
		expect(asFragment()).toMatchSnapshot();
	});
});
