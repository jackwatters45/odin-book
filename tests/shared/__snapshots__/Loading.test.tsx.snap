// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Loading Component > matches the snapshot 1`] = `
<DocumentFragment>
  .c0 {
  position: absolute;
  margin: auto;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  border: 12.5px solid #f3f3f3;
  border-top: 12.5px solid rgba(55, 55, 55, 0.5);
  border-radius: 50%;
  width: 50px;
  height: 50px;
  animation: dGOzbh 1.5s linear infinite;
}

<div
    class="c0"
    size="50"
  />
</DocumentFragment>
`;
